SimplenoteUITests/SimplenoteUITestsHistory.swift

```diff
@@ -0,0 +1,231 @@
+import XCTest
+
+class SimplenoteUISmokeTestsHistory: XCTestCase {
+    let notePrefix = "HistoryTest"
+
+    override class func setUp() {
+        app.launch()
+        getToAllNotes()
+        let _ = attemptLogOut()
+        EmailLogin.open()
+        EmailLogin.logIn()
+    }
+
+    override func setUpWithError() throws {
+        getToAllNotes()
+        NoteList.trashAllNotes()
+        Trash.empty()
+        NoteList.openAllNotes()
+    }
+
+    override class func tearDown() {
+        NoteList.trashAllNotes()
+        Trash.empty()
+    }
```

Comment:
This instance of teardown will be called only once at the end of the whole suit. 

Currently, all the tests in this file have a sort of tearDown at their end, which is usually closing History, leaving the editor, and trashing the note. 

But, at the same time (minus the history close part), the same is done in `setUpWithError`. Because of this, I would see value in having the following as the only step of `tearDown()` (note how you can change the teaDown declaration to have it executed at the end of _each_ test):

```suggestion
    override func tearDownWithError() throws {
        History.close()
    }
```

By doing so, you can remove the 'tearDown' part in all tests.

Reply:
I thought about using the `tearDown()` running after each test to have them more isolated, however part of what I wanted to do in `tearDown()` was to delete the note created during the test but not impacting a possible parallel execution (which might be out of the scope in this project), that's why I started using ids for each note. As I couldn't find a way to pass the name of the note to be deleted in `tearDown()` I moved to having it running once after the whole class is executed and getting rid of all the remaining notes.

I like the idea of not having tearDown code within the test method, what I miss in your suggestion is actually deleting the note created during the test. I can think in two solutions:
Assumption: parallelism is not a reality
1. We can go with your suggestion but keeping an after class `tearDown()` that would actually delete all the notes.
2. Add to your suggestion extra steps to delete all the notes after each tests.

I don't really like the option 2 because deleting all notes after each test feels too costly since parallelism is already out of scope.

What are your thoughts on this?

Reply:
The notes are already getting deleted as a part of `setUpWithError()`, which is executed before each test, this is why deleting them additionally at the end of the test or inside a tearDown means spending extra time 🤔 

Parallel execution under the same account is not the case: we have a dedicated solution for that on our CI, which manages a number of tests accounts and makes sure no account is used in parallel. While there is still a possibility for someone to run these tests locally using the test account which is hardcoded in the workspace, the chances and impact are so low, that we decided to neglect it.

The approach with using unique IDs for notes titles was one of the options we considered when we were solving the test accounts race condition, but a lot of existing tests rely on checking the total number of notes in `Trash` / `All Notes` etc, besides checking for the note content. Using this approach instead of having multiple accounts would mean loosing this possibility, as the number of notes cannot be predicted in this case (and this is the other reason why all notes are deleted at the beginning of each test).

In your case, using IDs in titles is a robust approach to check that `History` really reverts the revision 👍 . You are right about the fact that deleting all notes at the beginning of each test (instead of doing this only once at the beginning or end of the suite) is a bit of overkill, and you are free to adopt a different approach in the tests you've added. However, doing so for the other existing tests will cause fails, since they are written assuming that there are zero notes at the beginning of each test (except for `Search` tests).

The content for setUp/tearDown differs between the files with tests, and actually any form of tearDown is absent in the most files. You are free to have it defined the way you consider the best 🙂

Reply:
You're right, it makes total sense. I was kind of ignoring the `tearDown()` and `setup()` overlap.  I'll apply the changes. =)